src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+
+        return (int) $results->fetch()['count'];
+    }
+
+    public function getIterator() : Traversable
+    {
+        while (true) {
+            /** @var Statement $statement */
+            $results = $this->connection->executeQuery(
+                'SELECT version.id, version.version, version.content, version.hash, item.sequence'
+                ." FROM {$this->tableItems} item LEFT JOIN 
+{$this->tableVersions} version ON item.id = version.id AND item.version = version.version AND item.sequence > :sequence"
+                .' ORDER BY item.sequence LIMIT 100',
+                ['sequence' => $cursor ?? 0]
+            );
+
+            while ($result = $results->fetch(FetchMode::ASSOCIATIVE)) {
+                yield $this->rowToItemVersion($result);
+                $cursor = $result['sequence'];
+            }
+
+            if (empty($result)) {
+                break;
+            }
+        }
+    }
+
+    public function getSchema() : Schema
+    {
+        $schema = new Schema();
+
+        $table = $schema->createTable($this->tableItems);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('sequence', Type::INTEGER, ['autoincrement' => true]);
+        $table->setPrimaryKey(['sequence']);
+        $table->addUniqueIndex(['id']);
+
+        $table = $schema->createTable($this->tableVersions);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('content', Type::BLOB);
+        $table->addColumn('hash', Type::STRING);
+        $table->setPrimaryKey(['id', 'version']);
+        $table->addIndex(['id']);
+
+        return $schema;
+    }
+
+    private function rowToItemVersion(array $row) : ItemVersion
+    {
+        ['id' => $id, 'version' => $version, 'content' => $content, 'hash' => $hash] = $row;
+
+        // Some drivers, eg pdo_sqlite, unfortunately return strings rather than resources.
+        // Doctrine doesn't provide an equivalent of `PDOStatement::bindColumn()` either.
+        // This might cause memory problems if the content is too large.
+        $content = Type::getType('blob')->convertToPHPValue($content, $this->connection->getDatabasePlatform());
+
+        return new ItemVersion(ItemId::fromString($id), ItemVersionNumber::create($version), $content, $hash);
+    }
+
+    private function getCurrentVersion(ItemId $id, int $lockMode = LockMode::NONE) : ItemVersionNumber
+    {
+        $platform = $this->connection->getDatabasePlatform();
+
+        $sql = 'SELECT item.version FROM %s WHERE item.id = :id %s';
+        if (LockMode::NONE === $lockMode) {
+            $sql = sprintf($sql, "{$this->tableItems} item", '');
+        } else {
+            $sql = sprintf(
+                $sql,
+                $platform->appendLockHint("{$this->tableItems} item", LockMode::PESSIMISTIC_READ),
```

Comment:
This hardcodes rather than using `$lockMode`, but besides no pessimistic lock should be necessary for a content store that is dealing with low level of writes (seems to optimize reads) that are usually non-concurrent and indicate something's wrong then they are concurrent. Pessimistic writes in particular seem to block GETs while a new version is added, which I don't think we want to happen.

We should check the transaction(s) isolation level to see if there's any anomaly that could cause problems, like a dirty read of a version being added.

Reply:
`PESSIMISTIC_READ` should probably be used in these cases. But will need some proper testing.

Reply:
Commented "upstairs" on the transaction isolation levels, don't expect problems there too.